File: 1001-1100/1065-Index Pairs of a String/1065-Index Pairs of a String.py
import collections
class TrieNode:
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.isEnd = False

class Solution:
    def indexPairs(self, text: str, words: List[str]) -> List[List[int]]:
        root = TrieNode()
        for word in words:
            curr = root
            for c in word:
                curr = curr.children[c]
            curr.isEnd = True
            
        result = []
        for i in range(len(text)):
            j = i
            curr = root
            while j < len(text) and text[j] in curr.children:
                curr = curr.children[text[j]]
                if curr.isEnd:
                    result.append([i, j])
                j += 1
        return result

File: 1001-1100/1065-Index Pairs of a String/1065-Index Pairs of a String.cpp
class TrieNode {
public:
    TrieNode() {
        memset(next, NULL, sizeof(next));
        isEnd = false;
    }
    
    ~TrieNode() {
        for (TrieNode* p : next) {
            if (p) {
                delete p;
            }
        }
    }
    TrieNode *next[26];
    bool isEnd;
};

class Solution {
public:
    vector<vector<int>> indexPairs(string text, vector<string>& words) {
        TrieNode* root = new TrieNode();
        for (string word : words) {
            TrieNode* curr = root;
            for (char c : word) {
                if (curr->next[c - 'a'] == nullptr) {
                    curr->next[c - 'a'] = new TrieNode();
                }
                curr = curr->next[c - 'a'];
            }
            curr->isEnd = true;
        }
        
        vector<vector<int>> result;
        for (int i = 0; i < text.size(); ++i) {
            int j = i;
            TrieNode* curr = root;
            while (j < text.size() && curr->next[text[j] - 'a']) {
                curr = curr->next[text[j] - 'a'];
                if (curr->isEnd) {
                    result.push_back({i, j});
                }
                ++j;
            }
        }
        
        delete root;
        return result;
    }
};
